libsql/src/local/database.rs

```diff
@@ -351,6 +377,76 @@ impl Database {
         }
     }
 
+    #[cfg(feature = "sync")]
+    /// Push WAL frames to remote.
+    pub async fn push(&self) -> Result<crate::database::Replicated> {
+        let sync_ctx = self.sync_ctx.as_ref().unwrap();
+        let conn = self.connect()?;
+
+        let page_size = {
+            let rows = conn.query("PRAGMA page_size", crate::params::Params::None)?.unwrap();
+            let row = rows.next()?.unwrap();
+            let page_size = row.get::<u32>(0)?;
+            page_size
+        };
+
+        let mut max_frame_no: std::os::raw::c_uint = 0;
+        unsafe { libsql_sys::ffi::libsql_wal_frame_count(conn.handle(), &mut max_frame_no) };
+        
+        let generation = 1; // TODO: Probe from WAL.
+        let start_frame_no = sync_ctx.durable_frame_num + 1;
+        let end_frame_no = max_frame_no;
+
+        for frame_no in start_frame_no..end_frame_no+1 {
+            self.push_one_frame(&conn, &sync_ctx, generation, frame_no, page_size).await?;
+        }
+
+        let frame_count = end_frame_no - start_frame_no + 1;
+        Ok(crate::database::Replicated{
+            frame_no: None,
+            frames_synced: frame_count as usize,
+        })
+    }
+
+    #[cfg(feature = "sync")]
+    async fn push_one_frame(&self, conn: &Connection, sync_ctx: &SyncContext, generation: u32, frame_no: u32, page_size: u32) -> Result<()> {
+        let frame_size: usize = 24+page_size as usize;
+        let frame = vec![0; frame_size];
+        let rc = unsafe {
+            libsql_sys::ffi::libsql_wal_get_frame(conn.handle(), frame_no, frame.as_ptr() as *mut _, frame_size as u32)
+        };
+        if rc != 0 {
+            return Err(crate::errors::Error::SqliteFailure(rc as std::ffi::c_int, format!("Failed to get frame: {}", frame_no)));
+        }
+        let uri = format!("{}/sync/{}/{}/{}", sync_ctx.sync_url, generation, frame_no, frame_no+1);
+        self.push_with_retry(uri, &sync_ctx.auth_token, frame.to_vec(), sync_ctx.max_retries).await?;
+        Ok(())
+    }
+
+    #[cfg(feature = "sync")]
+    async fn push_with_retry(&self, uri: String, auth_token: &Option<String>, frame: Vec<u8>, max_retries: usize) -> Result<()> {
+        let mut nr_retries = 0;
+        loop {
+            let client = reqwest::Client::new();
```

Comment:
Whats the reason for using reqwest instead of just hyper directly? We also probably want to cache this client so we can re-use its dns resolution/tls handshake.